File: core/src/com/g3ida/withflyingcolours/core/player/movement/PlayerJumpSystem.java
package com.g3ida.withflyingcolours.core.player.movement;

import com.badlogic.ashley.core.Entity;
import com.badlogic.ashley.core.Family;
import com.badlogic.ashley.systems.IteratingSystem;
import com.badlogic.gdx.math.Vector2;
import com.g3ida.withflyingcolours.Utils;

import games.rednblack.editor.renderer.components.physics.PhysicsBodyComponent;
import games.rednblack.editor.renderer.utils.ComponentRetriever;

public class PlayerJumpSystem extends IteratingSystem {

    public PlayerJumpSystem() {
        super(Family.all(PhysicsBodyComponent.class, PlayerJumpComponent.class).get());
    }

    @Override
    protected void processEntity(Entity entity, float deltaTime) {
        PhysicsBodyComponent physicsBody = ComponentRetriever.get(entity, PhysicsBodyComponent.class);
        PlayerJumpComponent playerJump = ComponentRetriever.get(entity, PlayerJumpComponent.class);

        if(isGrounded(physicsBody)) {
            playerJump.timeSinceGrounded = 0;
        }

        boolean doJump = false; //should we jump in this frame ?

        if (isJumpPressed(playerJump)) {
            if (can_jump(playerJump, physicsBody)) {
                doJump = true;
            } else { // if the player can't jump right now let's check if he can before the responsiveness time passes.
                playerJump.responsivenessTimer = PlayerJumpComponent.jumpSettings.responsiveness;
            }
        }

        //handle responsiveness
        if (playerJump.responsivenessTimer > 0) {
            if (can_jump(playerJump, physicsBody)) {
                playerJump.responsivenessTimer = 0;
                doJump = true;
            } else {
                playerJump.responsivenessTimer -= deltaTime;
            }
        } else {
            playerJump.responsivenessTimer = 0;
        }

        //perform the jump
        if (doJump) {
            {
                playerJump.jumpTimer = playerJump.timeUntilFullJumpIsConsidered;
                physicsBody.body.applyLinearImpulse(0f, playerJump.jumpForce, physicsBody.body.getWorldCenter().x, physicsBody.body.getWorldCenter().y, true);
            }
        }

        //jump has been released before full jump reached
        if (!playerJump.shouldJump && playerJump.jumpTimer > Utils.epsilon) {
            //cancel jump
            Vector2 velocity = physicsBody.body.getLinearVelocity();
            if (velocity.y > 0) { // decrease velocity only if the player is going up !
                physicsBody.body.setLinearVelocity(velocity.x, velocity.y * 0.5f);
            }
        }

        if (playerJump.shouldJump) {
            playerJump.jumpTimer -= deltaTime;
        }

        if(!isGrounded(physicsBody)) {
            playerJump.timeSinceGrounded += deltaTime;
        }

        playerJump.oldShouldJump = playerJump.shouldJump;
    }

    boolean isJumpPressed(PlayerJumpComponent playerJump) {
        return !playerJump.oldShouldJump && playerJump.shouldJump;
    }

    boolean isJumpReleased(PlayerJumpComponent playerJump) {
        return playerJump.oldShouldJump && !playerJump.shouldJump;
    }

    boolean isJumpDown(PlayerJumpComponent playerJump) {
        return playerJump.shouldJump;
    }

    boolean isGrounded(PhysicsBodyComponent physicsBody) {
        // FIXME: this condition is also met on jump peak.
        return Math.abs(physicsBody.body.getLinearVelocity().y) < Utils.epsilon;
    }

    boolean can_jump(PlayerJumpComponent playerJump, PhysicsBodyComponent physicsBody) {
        if(isGrounded(physicsBody)) {
            return true;
        }
        //handle permissiveness
        boolean isJumping = physicsBody.body.getLinearVelocity().y > 0f;
        return (playerJump.timeSinceGrounded <= PlayerJumpComponent.jumpSettings.permissiveness) && !isJumping && playerJump.jumpTimer <= 0f;
    }
}

File: core/src/com/g3ida/withflyingcolours/core/player/movement/PlayerJumpSystem.kt
package com.g3ida.withflyingcolours.core.player.movement

import com.artemis.ComponentMapper
import com.artemis.systems.IteratingSystem
import games.rednblack.editor.renderer.components.physics.PhysicsBodyComponent
import com.artemis.annotations.All
import com.g3ida.withflyingcolours.Utils

@All(PhysicsBodyComponent::class, PlayerJumpComponent::class)
class PlayerJumpSystem : IteratingSystem() {
    var mPhysicsBodyComponent: ComponentMapper<PhysicsBodyComponent>? = null
    var mPlayerJumpComponent: ComponentMapper<PlayerJumpComponent>? = null
    override fun process(entityId: Int) {
        val physicsBody = mPhysicsBodyComponent!![entityId]
        val playerJump = mPlayerJumpComponent!![entityId]
        if (isGrounded(physicsBody)) {
            playerJump.timeSinceGrounded = 0f
        }
        var doJump = false //should we jump in this frame ?
        if (isJumpPressed(playerJump)) {
            if (can_jump(playerJump, physicsBody)) {
                doJump = true
            } else { // if the player can't jump right now let's check if he can before the responsiveness time passes.
                playerJump.responsivenessTimer = PlayerJumpComponent.jumpSettings.responsiveness
            }
        }

        //handle responsiveness
        if (playerJump.responsivenessTimer > 0) {
            if (can_jump(playerJump, physicsBody)) {
                playerJump.responsivenessTimer = 0f
                doJump = true
            } else {
                playerJump.responsivenessTimer -= world.delta
            }
        } else {
            playerJump.responsivenessTimer = 0f
        }

        //perform the jump
        if (doJump) {
            run {
                playerJump.jumpTimer = playerJump.timeUntilFullJumpIsConsidered
                physicsBody.body.applyLinearImpulse(0f, playerJump.jumpForce, physicsBody.body.worldCenter.x, physicsBody.body.worldCenter.y, true)
            }
        }

        //jump has been released before full jump reached
        if (!playerJump.shouldJump && playerJump.jumpTimer > Utils.EPSILON) {
            //cancel jump
            val velocity = physicsBody.body.linearVelocity
            if (velocity.y > 0) { // decrease velocity only if the player is going up !
                physicsBody.body.setLinearVelocity(velocity.x, velocity.y * 0.5f)
            }
        }
        if (playerJump.shouldJump) {
            playerJump.jumpTimer -= world.delta
        }
        if (!isGrounded(physicsBody)) {
            playerJump.timeSinceGrounded += world.delta
        }
        playerJump.oldShouldJump = playerJump.shouldJump
    }

    fun isJumpPressed(playerJump: PlayerJumpComponent): Boolean {
        return !playerJump.oldShouldJump && playerJump.shouldJump
    }

    fun isJumpReleased(playerJump: PlayerJumpComponent): Boolean {
        return playerJump.oldShouldJump && !playerJump.shouldJump
    }

    fun isJumpDown(playerJump: PlayerJumpComponent): Boolean {
        return playerJump.shouldJump
    }

    fun isGrounded(physicsBody: PhysicsBodyComponent): Boolean {
        // FIXME: this condition is also met on jump peak.
        return Math.abs(physicsBody.body.linearVelocity.y) < Utils.EPSILON
    }

    fun can_jump(playerJump: PlayerJumpComponent, physicsBody: PhysicsBodyComponent): Boolean {
        if (isGrounded(physicsBody)) {
            return true
        }
        //handle permissiveness
        val isJumping = physicsBody.body.linearVelocity.y > 0f
        return (playerJump.timeSinceGrounded <= PlayerJumpComponent.jumpSettings.permissiveness
                && !isJumping && playerJump.jumpTimer <= 0f)
    }
}